lib/logstash/filters/geoip/database_manager.rb

```diff
@@ -0,0 +1,273 @@
+      if execute_download_check
+        @geoip.setup_filter(database_path)
+      end
+    rescue DatabaseExpiryError => e
+      logger.error(e.message, :cause => e.cause, :backtrace => e.backtrace)
+      @geoip.terminate_filter
+    end
+  end
+
+  def close
+    @scheduler.every_jobs.each(&:unschedule) if @scheduler
+  end
+
+  # provide backward compatibility for LS < 7.12 which is always in offline mode
+  def get_mode(database_path)
+    (database_path.nil? and logstash_version >= 7.12)? :online : :offline
+  end
+
+  def database_path
+    @database_path
+  end
+
+  protected
+  # Resolve database path from metadata and validate with md5
+  # Write current timestamp if metadata file is missing
+  def setup_metadata
+    metadata = get_metadata.last
+
+    if metadata
+      path = get_file_path(metadata[Column::FILENAME])
+      if file_exist?(path) && (md5(path) == metadata[Column::MD5])
+        @database_path = path
+      end
+    else
+      save_timestamp
+    end
+  end
+
+  # return a valid database path or default database path
+  def patch_database_path(database_path)
+    unless file_exist?(database_path)
+      database_path = get_file_path("GeoLite2-#{@database_type}.mmdb")
+
+      unless file_exist?(database_path)
+        raise "You must specify 'database => ...' in your geoip filter (I looked for '#{database_path}')"
+      end
+    end
+
+    database_path
+  end
+
+  # csv format: database_type, update_at, gz_md5, md5, filename
+  def save_timestamp
+    metadata = get_metadata(false)
+    metadata << [@database_type, Time.now.to_i, md5(database_zip_path), md5(@database_path), database_filename]
+
+    ::CSV.open @metadata_path, 'w' do |csv|
+      metadata.each { |row| csv << row }
+    end
+  end
+
+  # Call infra endpoint to get md5 of latest database and verify with metadata
+  # return [has_update, server db info]
+  def check_update
+    uuid = get_uuid
+    res = rest_client.get("#{GEOIP_ENDPOINT}?key=#{uuid}")
+
+    all_db = JSON.parse(res.body)
+    target_db = all_db.select { |info| info['name'].include?(@database_type) }.first
+
+    metadata = get_metadata.last
+    if metadata
+      [metadata[Column::GZ_MD5] != target_db['md5_hash'], target_db]
+    else
+      [true, target_db]
+    end
+  end
+
+  def download_database(server_db)
+    Stud.try(3.times) do
+      new_database_zip_path = get_file_path(server_db['name'].gsub(@database_type, "#{@database_type}_#{Time.now.to_i}"))
+      Down.download(server_db['url'], destination: new_database_zip_path)
+      raise "the new download has wrong checksum" if md5(new_database_zip_path) != server_db['md5_hash']
+      new_database_zip_path
+    end
+  end
+
+  def unzip(zip_path)
+    database_path = zip_path[0...-3]
+    Zlib::GzipReader.open(zip_path) do |gz|
+      ::File.open(database_path, "wb") do |f|
+        f.print gz.read
+      end
+    end
+    database_path
+  end
+
+  # Make sure the path has usable database
+  def assert_database(database_path)
```

Comment:
It raises an exception if something is not ok, so probably `assert_database!` is best